R/060_utils.R

```diff
@@ -0,0 +1,348 @@
+#'
+#' @param header_df a data frame containing header cells.
+#' @param header_fill he criteria according to which blank cells will be filled.
+#' @param formats the formats of the workbook associated with the tidyxl data frame.
+#' @param direction the direction in which headers are filled.
+
+fill_blanks_in_headers <- function(header_df, header_fill, formats, direction) {
+  if (direction %in% c("N", "S", "up", "down")) {
+    header_df <- suppressMessages(fill_blanks_in_col_headers(header_df, header_fill, formats))
+  } else if (direction %in% c("W", "E", "left", "right")) {
+    header_df <- suppressMessages(fill_blanks_in_row_headers(header_df, header_fill, formats))
+  } else {
+    stop("Please check the direction you are providing")
+  }
+
+  header_df
+}
+
+#' Add a set of variables based on a list of formulae
+#'
+#' This is an internal function that takes a data frame and a list of formulae and appends an arbitrary number of columns to the data frame based on the formulae.
+#'
+#' @param df a data frame containing header cells.
+#' @param form_list list of formulae
+#' @param format the formats associated with the workbook containing the header_df cells.
+#'
+
+reduce_mutated <- function(df, form_list, format) {
+  current_quosure <- form_list[[1]]
+  var_name_sym <- rlang::sym(form_list[[2]])
+
+  df %>%
+    dplyr::mutate(!!var_name_sym := !!current_quosure)
+}
+
+#' Give quosure a name
+#'
+#' This is an internal function that adds a prefixed name to a quosure so that variables added to a data frame using this quosure have a predictable name.
+#' @param x quosure
+#' @param prefix prefix to be added in name
+#'
+
+append_name_to_quosure <- function(x, prefix = "grp_") {
+  list(
+    x,
+    paste0(
+      prefix,
+      x %>% rlang::as_label() %>% make.names() %>%
+        stringr::str_replace_all("\\.+", ".") %>% stringr::str_remove_all("(\\.$)|(^X\\.)") %>%
+        stringr::str_replace_all("\\.", "_") %>%
+        ifelse(stringr::str_sub(., start = 1, 1) %in% as.character(0:9), paste0("x", .), .)
+    )
+  )
+}
+
+#' Add a variable to a tidyxl data frame if that variable is missing
+#'
+#' This is an internal function that takes a tidyxl data frame and a variable name to be added if they are missing.
+#' @param sheet sheet nominated for tidying
+#' @param var name of the var to be added as a character string.
+#'
+add_variable_if_missing <- function(sheet, var) {
+  if (!var %in% names(sheet)) {
+    var_sym <- rlang::sym(var)
+
+    sheet <- sheet %>% dplyr::mutate(!!var_sym := NA_character_)
+  }
+
+  sheet
+}
+
+#' Create a filtering variable based on a excel string range.
+#'
+#' This is an internal function that create a variabled prefixed "flt" based on whether cells are within a range specified by an excel string range.
+#' @param table_range a table_range given by string. For example, "A1:B2"
+#' @param sheet a tidyxl data frame.
+
+string_range_to_filter_vars <- function(sheet, table_range) {
```

Comment:
I don't follow the process in `locate_data.R` that uses this function.  It seems to programmatically create functions with an `flt_` prefix, and then apply them in turn to the `local_format_id` column.  Could there be another way to do whatever is being done?

Reply:
That's right. It create's a set of variables with an flt_ prefix. These variables are then used to filter the data frame. It might be better to avoid storing the functions in the environment and instead put them in a list. But that is only slightly different. I'll need to give this some more thought.

Reply:
I need to tidy this up, but I've changed to a much cleaner approach in this commit: 90a55d5bea61822baf313057453defc2d53b9943. It uses named quosures along the lines of https://github.com/nacnudus/unpivotr/pull/33#discussion_r365459639.

Reply:
Tidied up approach above with this commit: fa6bef1c4ddc11f61a8e4bf68aca98eaaac77333.